cpp/ql/lib/ext/std.array.model.yml

```diff
@@ -0,0 +1,12 @@
+extensions:
+  - addsTo:
+      pack: codeql/cpp-all
+      extensible: summaryModel
+    data: # namespace, type, subtypes, name, signature, ext, input, output, kind, provenance
+      - ["std", "array", True, "at", "", "", "Argument[-1].Element[*@]", "ReturnValue[*@]", "value", "manual"]
```

Comment:
Why are all the models using `Element[*]` (or `Element[*@]`) and not just `Element[]` (or `Element[@]`)?  Surely `array.at(0)` contains elements, not element references?

Reply:
Both `vector::at` and `vector::operator[]` return references to their elements: https://en.cppreference.com/w/cpp/container/vector/at

This is why you can do stuff like:
```cpp
std::vector<int> v{10};
v.at(0) = 42;
```

Reply:
Yep, the return value is a reference, but the input `Argument[-1].Element[*@]` surprised me.  Am I reading this right as "one or more dereferences of an element contained in the qualifier"?  What are the dereferences about?

Reply:
Because you can also do:
```cpp
std::vector<int*> v{10};
// ...
*v.at(0) = x;
```
That is, it's not the content that's tainted, it's the value of the pointed to content.

Reply:
Or equivalently, you can do:
```cpp
std::vector<int*> v{10};
int x = source();
v.at(0) = &x;
```

Reply:
You are absolutely right. In https://github.com/github/codeql/pull/16791/commits/5be948533c22525716758721635fe69095c11492 I've replaced all mentions of `Element[*@]` with `Element[@]`.